nixos/modules/services/networking/pangolin.nix

```
@@ -0,0 +1,590 @@
+{
+  utils,
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.pangolin;
+  format = pkgs.formats.yaml { };
+  finalSettings = lib.attrsets.recursiveUpdate pangolinConf cfg.settings;
+  cfgFile = format.generate "config.yml" finalSettings;
+
+  pangolinConf = {
+    app = {
+      dashboard_url = "https://${cfg.dashboardDomain}";
+      log_level = "info";
+      save_logs = false;
+    };
+    domains.domain1 = {
+      base_domain = cfg.baseDomain;
+      cert_resolver = "letsencrypt";
+      prefer_wildcard_cert = false;
+    };
+    server = {
+      external_port = 3000;
+      internal_port = 3001;
+      next_port = 3002;
+      internal_hostname = "localhost";
+      session_cookie_name = "p_session_token";
+      resource_access_token_param = "p_token";
+      resource_access_token_headers = {
+        id = "P-Access-Token-Id";
+        token = "P-Access-Token";
+      };
+      resource_session_request_param = "p_session_request";
+    };
+    traefik = {
+      cert_resolver = "letsencrypt";
+      http_entrypoint = "web";
+      https_entrypoint = "websecure";
+    };
+    gerbil = {
+      start_port = 51820;
+      base_endpoint = cfg.dashboardDomain;
+      block_size = 24;
+      site_block_size = 30;
+      subnet_group = "100.89.128.1/24";
+      use_subdomain = false;
+    };
+    rate_limits.global = {
+      window_minutes = 1;
+      max_requests = 500;
+    };
+    users.server_admin = {
+      email = cfg.letsEncryptEmail;
+      password = "Password123!"; # this is a dummy password and is overwritten by the environment variable USERS_SERVERADMIN_PASSWORD declared in pangolinEnvironmentFile
+    };
+  };
+in
+{
+  options = {
+    services.pangolin = {
+      enable = lib.mkEnableOption "Pangolin";
+      package = lib.mkPackageOption pkgs "fosrl-pangolin" { };
+
+      settings = lib.mkOption {
+        inherit (format) type;
+        default = { };
+        description = ''
+          Additional attributes to be merged with the configuration options and written to Pangolin's config.yml file.
+        '';
+        example = {
+          app = {
+            save_logs = true;
+          };
+          server = {
+            external_port = 3007;
+            internal_port = 3008;
+          };
+          domains.domain1 = {
+            prefer_wildcard_cert = true;
+          };
+        };
+      };
+      openFirewall = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Whether to open TCP ports 80 and 443, and UDP port 51820 in the firewall for the Pangolin service(s).
+        '';
+      };
+      baseDomain = lib.mkOption {
+        type = with lib.types; nullOr str;
+        default = null;
+        description = ''
+          Your base fully qualified domain name (without any subdomains).
+        '';
+        example = "example.com";
+      };
+      dashboardDomain = lib.mkOption {
```

```suggestion
      };

      dashboardDomain = lib.mkOption {
```
Nit, add a blank line between each option declaration.